GH-35419: [GLib] Add GArrowFixedShapeTensorDataType

hiroyuki-sato opened the pull request:
### Rationale for this change

The C++ API implemented FixedShapeTensor Extension DataType. GLib was not yet supported.

### What changes are included in this PR?

Implement GArrowFixedShapeTensorDataType.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* GitHub Issue: #35419

hiroyuki-sato commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2269,37 @@ garrow_string_view_data_type_new(void)
+                 NULL));
+  return data_type;
+}
+
```
@kou Should I move to outside of  `G_BEGIN_DECLS ` ~ `G_END_DECLS ` parts?

kou replied:
No.

We should place C functions in `G_BEGIN_DECLS`/`G_END_DECLS`.

hiroyuki-sato commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2269,37 @@ garrow_string_view_data_type_new(void)
+GArrowFixedShapeTensorDataType *
+garrow_fixed_shape_tensor_data_type_new(void)
+{
+  auto arrow_data_type = arrow::GetExtensionType("arrow.fixed_shape_tensor");
```
Could you check this part correct?

kou replied:
We need to accept `value_type`, shape`, `permutation` and `dimension_names` and use `std::make_shared<arrow::FixedShapeTensorType(...)` instead of `arrow::GetExtensionType(...)`.

See also: https://github.com/apache/arrow/blob/068416bd411d6a8e2949f8ebcb2f80e2c302ef6b/cpp/src/arrow/extension/fixed_shape_tensor.h#L55-L58

hiroyuki-sato replied:
`permutation` and `dim_names` allows empty. Do we need to add `(nullable)` annotation?

```diff
diff --git a/c_glib/arrow-glib/basic-data-type.cpp b/c_glib/arrow-glib/basic-data-type.cpp
index 2f547ee60..012562828 100644
--- a/c_glib/arrow-glib/basic-data-type.cpp
+++ b/c_glib/arrow-glib/basic-data-type.cpp
@@ -2291,11 +2291,11 @@ garrow_fixed_shape_tensor_data_type_class_init(GArrowFixedShapeTensorDataTypeCla
  * @shape: (array length=shape_length): A physical shape of the contained tensors as an
  * array.
  * @shape_length: The length of `shape`.
- * @permutation: (array length=permutation_length): An indices of the desired ordering of
+ * @permutation: (array length=permutation_length) (nullable): An indices of the desired ordering of
  * the original dimensions, defined as an array. This must be `NULL` or the same length
  * array of `shape`.
  * @permutation_length: The length of `permutation`.
- * @dim_names: (array length=n_dim_names): Explicit names to tensor dimensions as an
+ * @dim_names: (array length=n_dim_names) (nullable): Explicit names to tensor dimensions as an
  * array. This must be `NULL` or the same length array of `shape`.
  * @n_dim_names. The length of `dim_names`.
  * @error: (nullable): Return location for a #GError or %NULL.
```

kou replied:
I'm not sure `(nullable)` works here. Let's try it.

hiroyuki-sato replied:
I added `(nullable)` annotation and tests.

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,73 @@ garrow_string_view_data_type_new(void)
+
+/**
+ * garrow_fixed_shape_tensor_data_type_new:
+ * @value_type: A #GArrowDataType.
```
```suggestion
 * @value_type: A #GArrowDataType of individual tensor elements.
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,73 @@ garrow_string_view_data_type_new(void)
+/**
+ * garrow_fixed_shape_tensor_data_type_new:
+ * @value_type: A #GArrowDataType.
+ * @size: The length of `shape`.
+ * @psize: The length of `permutation`.
+ * @shape: (array length=size): shape.
+ * @permutation: (array length=psize): permutation.
+ * @dim_name: (array length=size): dim_names.
```
`permutation` and `dim_names` can be `NULL`. So we can't share the same length argument.

```suggestion
 * @shape: (array length=shape_length): A physical shape of the contained tensors as an array.
 * @shape_length: The length of `shape`.
 * @permutation: (array length=permutation_length): An indices of the desired ordering of the original
 *    dimensions, defined as an array. This must be `NULL` or the same length array of `shape`.
 * @permutation_length: The length of `permutation`.
 * @dim_names: (array length=n_dim_names): Explicit names to tensor dimensions as an array.
 *   This must be `NULL` or the same length array of `shape`.
 * @n_dim_names. The length of `dim_names`.
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,73 @@ garrow_string_view_data_type_new(void)
+  auto arrow_value_type = garrow_data_type_get_raw(value_type);
+
+  for (int i = 0; i < psize; i++) {
+    arrow_shape.push_back(permutation[i]);
```
`arrow_permutation`?

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,73 @@ garrow_string_view_data_type_new(void)
+    arrow_dim_names.push_back(dim_name[i]);
+  }
+
+  auto arrow_data_type = arrow::extension::FixedShapeTensorType::Make(arrow_value_type,
+                                                                      arrow_shape,
+                                                                      arrow_permutation,
+                                                                      arrow_dim_names);
+
+  if (!garrow::check(error, arrow_data_type, "[fixed-shape-tensor][new]")) {
```
```suggestion
  auto arrow_data_type_result = arrow::extension::FixedShapeTensorType::Make(arrow_value_type,
                                                                      arrow_shape,
                                                                      arrow_permutation,
                                                                      arrow_dim_names);
  if (!garrow::check(error, arrow_data_type_result, "[fixed-shape-tensor][new]")) {
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,73 @@ garrow_string_view_data_type_new(void)
+  }
+
+  GArrowFixedShapeTensorDataType *data_type = GARROW_FIXED_SHAPE_TENSOR_DATA_TYPE(
+    g_object_new(GARROW_TYPE_FIXED_SHAPE_TENSOR_DATA_TYPE,
+                 "data-type",
+                 &arrow_data_type.ValueUnsafe(),
```
```suggestion
    auto arrow_data_type = *arrow_data_type_result;
    g_object_new(GARROW_TYPE_FIXED_SHAPE_TENSOR_DATA_TYPE,
                 "data-type",
                 &arrow_data_type,
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,73 @@ garrow_string_view_data_type_new(void)
+    return NULL;
+  }
+
+  GArrowFixedShapeTensorDataType *data_type = GARROW_FIXED_SHAPE_TENSOR_DATA_TYPE(
```
```suggestion
  auto data_type = GARROW_FIXED_SHAPE_TENSOR_DATA_TYPE(
```

hiroyuki-sato commented:
@kou Thanks! Fixed review comments. Please take a look when you get a chance. (I have a question).

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,83 @@ garrow_string_view_data_type_new(void)
+ * garrow_fixed_shape_tensor_data_type_new:
+ * @value_type: A #GArrowDataType of individual tensor elements.
+ * @shape: (array length=shape_length): A physical shape of the contained tensors as an
+ * array.
```
```suggestion
 *   array.
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,83 @@ garrow_string_view_data_type_new(void)
+ * array.
+ * @shape_length: The length of `shape`.
+ * @permutation: (array length=permutation_length): An indices of the desired ordering of
+ * the original dimensions, defined as an array. This must be `NULL` or the same length
+ * array of `shape`.
```
```suggestion
 *   the original dimensions, defined as an array. This must be `NULL` or the same length
 *   array of `shape`.
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,83 @@ garrow_string_view_data_type_new(void)
+ * array of `shape`.
+ * @permutation_length: The length of `permutation`.
+ * @dim_names: (array length=n_dim_names): Explicit names to tensor dimensions as an
+ * array. This must be `NULL` or the same length array of `shape`.
```
```suggestion
 *   array. This must be `NULL` or the same length array of `shape`.
```

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -2267,6 +2271,83 @@ garrow_string_view_data_type_new(void)
+GArrowFixedShapeTensorDataType *
+garrow_fixed_shape_tensor_data_type_new(GArrowDataType *value_type,
+                                        const gint64 *shape,
+                                        gint32 shape_length,
```
Could you use `gsize` like other functions?

```suggestion
                                        gsize shape_length,
```

kou commented on c_glib/test/test-fixed-shape-tensor-data-type.rb:
```diff
@@ -0,0 +1,82 @@
+    assert_equal(message,
+                 error.message.lines.first.chomp)
+  end
+
```
```suggestion
```

kou commented on c_glib/test/test-fixed-shape-tensor-data-type.rb:
```diff
@@ -0,0 +1,82 @@
+# under the License.
+
+class TestFixedShapeTensorDataType < Test::Unit::TestCase
+  def test_type
```
```suggestion
  def test_id
```

kou commented on c_glib/test/test-fixed-shape-tensor-data-type.rb:
```diff
@@ -0,0 +1,82 @@
+                                                    [3, 4],
+                                                    [1, 0],
+                                                    ["x", "y"])
+    assert_true(data_type.to_s.start_with?("extension<arrow.fixed_shape_tensor"))
```
Could you use power-assert for predicate?

```suggestion
    assert do
      data_type.to_s.start_with?("extension<arrow.fixed_shape_tensor")
    end
```

kou commented on c_glib/test/test-fixed-shape-tensor-data-type.rb:
```diff
@@ -0,0 +1,82 @@
+                                                    [],
+                                                    [],
+                                                    [])
+    assert_equal(Arrow::Type::EXTENSION, data_type.id)
```
Can we change this when we add `Arrow::FixedShapeTensorDataType#shape` in follow-up PR?

hiroyuki-sato replied:
Removed this test. This test creates the following instance. I'm not sure whether  this instance is useful or not. 

```
#<Arrow::FixedShapeTensorDataType:0x00000001250349d8 ptr=0x000060000054fee0 extension<arrow.fixed_shape_tensor[value_type=uint64, shape=[]]>>
```

kou replied:
@rok Do you think that `Farrow::FixedShapeTensorType` rejects an empty shape because it's useless?

hiroyuki-sato commented:
@kou Please take a look when you get a chance.

* Applied your suggestion.
* Added `(nullable)` annotation `permutation` and `dim_names`.
* Removed `Arrow::FixedShapeTensorDataType.new(Arrow::UInt64DataType.new,[],[],[])` tests

kou commented on c_glib/arrow-glib/basic-data-type.cpp:
```diff
@@ -25,6 +25,8 @@
 #include <arrow-glib/field.hpp>
 #include <arrow-glib/type.hpp>
 
+#include <arrow/extension/fixed_shape_tensor.h>
+
 #include <arrow/c/bridge.h>
```
```suggestion
#include <arrow/c/bridge.h>
#include <arrow/extension/fixed_shape_tensor.h>
```

kou commented on c_glib/test/test-fixed-shape-tensor-data-type.rb:
```diff
@@ -0,0 +1,91 @@
+                                                    [3, 4],
+                                                    nil,
+                                                    ["x", "y"])
+    assert_equal(Arrow::Type::EXTENSION, data_type.id)
```
TODO: We should change this assertion when we add `Arrow::FixedShapeTensorDataType#permutation`.

hiroyuki-sato replied:
Added TODO: comment.

kou replied:
Could you do it for `nil` `dim_names` too?

hiroyuki-sato replied:
I also added TODO comment.

```diff
diff --git a/c_glib/test/test-fixed-shape-tensor-data-type.rb b/c_glib/test/test-fixed-shape-tensor-data-type.rb
index b88967acc..2e7bc2cbf 100644
--- a/c_glib/test/test-fixed-shape-tensor-data-type.rb
+++ b/c_glib/test/test-fixed-shape-tensor-data-type.rb
@@ -57,6 +57,7 @@ class TestFixedShapeTensorDataType < Test::Unit::TestCase
                                                     [3, 4],
                                                     [0, 1],
                                                     nil)
+    # TODO: Use Arrow::FixedShapeTensorDataType#dim_names
     assert_equal(Arrow::Type::EXTENSION, data_type.id)
   end
```

kou reviewed (approved):
+1
